torchtitan/utils.py

```diff
@@ -48,6 +48,10 @@ def set_determinism(seed: Optional[int]) -> None:
         torch.backends.cudnn.benchmark = False
         # set Python seed
         os.environ["PYTHONHASHSEED"] = str(seed)
+        torch.use_deterministic_algorithms(True)
```

Comment:
I was going to add this earlier but I found it crashes compile when used with fp8:
"torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
  RuntimeError: "fill_out" not implemented for 'Float8_e4m3fn'"
  
  Thus, I don't think we want to add this atm. 
  It works (compiles) if you don't use fp8 but a lot of the need for determinism is to better showcase fp8.

Reply:
I filed an issue to start work on getting this resolved:
https://github.com/pytorch/pytorch/issues/137160

Reply:
thanks for explaining the context

Reply:
We should leave a TODO here reminding us to use it after the issue gets fixed.

also tbh I don't know what exactly `torch.use_deterministic_algorithms` does. Does it cover everything else we are doing here?